### ISSUE TITLE
Corrupted NSEC3 RRs in AXFR responses

### ISSUE THREAD

**Issue by monodhs** [PowerDNS/pdns#10591]
- Program: Authoritative
 - Issue type: Defect report

### Short description
PowerDNS fails to produce faithful NSEC3 RRs in its response to AXFR requests when serving a "pre-signed" zone. The result of which are bogus signature errors from data served by my secondaries (and beyond) when attempting to validate the denial of existence of a RR. However, when issued the same request directly in its function as an authoritative server, the NSEC3 RR in question is served faithfully and no bogus signature errors are produced, and denial of existence is validated successfully.

### Environment

 - Operating system: Debian GNU/Linux 10 (buster)
 - Software version: 4.1.6
 - Software source: https://packages.debian.org/buster/pdns-server

### Steps to reproduce
I will gladly provide network access to the server in question to observe the phenomena described here. For that to work, however, I will need to know in advance the IP address from which the server is going to be accessed. I can also provide the complete, signed, and verified master zone file with which I load the server, if interested.

My OpenPGP key is available on any decent key server.

### Expected behaviour
Faithful reproduction of NSEC3 RRs from the master zone file when serving zone data in response to AXFR requests.

### Actual behaviour
In particular, it is the NSEC3 RR of the zone apex that has the DNSKEY bit set in the complete, signed, and verified master zone file, but that DNSKEY bit is cleared in the corresponding NSEC3 RR PowerDNS serves in its responses to AXFR requests.

### Other information

Thanks!

**Comment by rgacogne** [comment]
Thank you for reporting this! I'm afraid 4.1.6 is a very old version, that we don't support anymore. Would you be able to check whether the problem occurs on a newer version? We do provide packages for buster on our repository: https://repo.powerdns.com/

**Comment by klaus-nicat** [comment]
I can confirm the corruption with pdns 4.1.13. My setup: Bind (signer) -> PDNS -> other Secondaries: In may case it complains about the following 2 changes (comparing an AXFR from the Bind Primary against AXFR from PDNS):

```
-KURDOQ3EV59TT2J2VSR0PMUTUNNCMF35.nic.at. 900 IN        NSEC3 1 0 10 7312B0F1 LH5U7F525DLARTRBQRBU3TKN2N9B4T8I A NS SOA MX TXT RRSIG DNSKEY NSEC3PARAM TYPE65534
+kurdoq3ev59tt2j2vsr0pmutunncmf35.nic.at. 900 IN        NSEC3 1 0 10 7312B0F1 LH5U7F525DLARTRBQRBU3TKN2N9B4T8I A NS SOA MX TXT RRSIG NSEC3PARAM TYPE65534
```
PDNS omits the DNSKEY in the signature.

```
-nic.at.                        0       IN      NSEC3PARAM 1 0 10 7312B0F1
+nic.at.                        900     IN      NSEC3PARAM 1 0 10 7312B0F1
```

PDNS Sets the TTL of NSEC3PARAM to 900 instead of the original. I guess PDNS does not store the original TTL when converting the NSEC3PARAM to Domainmetadata.

I did not had a chance yet to test with 4.5.

**Comment by Habbie** [comment]
Of those differences, only the DNSKEY missing from the NSEC3 bitmap would cause trouble, I think.

**Comment by klaus-nicat** [comment]
We have not experienced any troubles yet (2 of 5 public facing secondaries omit the DNSKEY bit). I guess it only causes NXDOMAIN to fail which results in SERVFAIL instead of NXDOMAIN und would not cause troubles to end users.

**Comment by monodhs** [comment]
> ```
> -KURDOQ3EV59TT2J2VSR0PMUTUNNCMF35.nic.at. 900 IN        NSEC3 1 0 10 7312B0F1 LH5U7F525DLARTRBQRBU3TKN2N9B4T8I A NS SOA MX TXT RRSIG DNSKEY NSEC3PARAM TYPE65534
> +kurdoq3ev59tt2j2vsr0pmutunncmf35.nic.at. 900 IN        NSEC3 1 0 10 7312B0F1 LH5U7F525DLARTRBQRBU3TKN2N9B4T8I A NS SOA MX TXT RRSIG NSEC3PARAM TYPE65534
> ```

That's precisely what I am seeing.

My guess is (haven't looked at the code, yet), that PowerDNS produces the NSEC3 RR on-the-fly when responding to the AXFR request and forgets to set the DNSKEY bit in the NSEC3 RR. This can be easily observed in the AXFR response, where the NSEC3 RRs are trailing the other RRs, including RRSIG RRs for those NSEC2 RRs.

Due to the DNSKEY bit being omitted in the NSEC3 RR, and with no way to sign this particular NSEC3 RR, PowerDNS hands over its copy of the original RRSIG that was produced from the original NSEC3 RR elsewhere, wherein the DNSKEY bit **is** set. This ultimately leads to the vaiidating resolver claiming that the RRSIG in question is bogus, because it was produced from an ever so slightly different NSEC3 RR which has the DNSKEY bit set.

**Comment by pieterlexis** [comment]
can you provide your pdns.conf? I am very interested in the `direct-dnskey` setting (but also the rest)

**Comment by monodhs** [comment]
> Thank you for reporting this! I'm afraid 4.1.6 is a very old version, that we don't support anymore. Would you be able to check whether the problem occurs on a newer version?

Not quite yet. Hence the ancient version.

**Comment by klaus-nicat** [comment]
direct-dnskey=yes

**Comment by monodhs** [comment]
> can you provide your pdns.conf? I am very interested in the `direct-dnskey` setting (but also the rest)

With reference to [the documentation](https://doc.powerdns.com/authoritative/settings.html#direct-dnskey), I can say that `direct-dnskey` is set to `no`, simply because it's the first time I am made aware of its existence ;)

Which version was `direct-dnskey` introduced?

**Comment by klaus-nicat** [comment]
With direct-dnskey=no the DNSKEY bit is present

**Comment by monodhs** [comment]
> With direct-dnskey=no the DNSKEY bit is present

If it is, where does PowerDNS obtain the DNSKEY RRs for a zone in a pre-signed setup, when the master files are the **only** source of truth?

**Comment by pieterlexis** [comment]
DNSKEYs are in the zone (we don't need the private keys, as we use the existing RRSIGs in the pre-signed case)

**Comment by monodhs** [comment]
> DNSKEYs are in the zone (we don't need the private keys, as we use the existing RRSIGs in the pre-signed case)

Shall I just set `direct-dnskey=no`, then?

**Comment by klaus-nicat** [comment]
I just tested with pdns 4.5 and the DNSKEY bit is set in both cases - direct-dnskey=no and direct-dnskey=yes. So I guess this DNSKEY-bit bug is fixed in 4.5.

Regarding the wrong TTL of the NSEC3PARAM RR: dnssec-verify complains, ldns-verify-zone does not complain.

**Comment by Habbie** [comment]
The wrong TTL for NSEC3PARAM is tracked in #2304 but we have no plans to fix it.

**Comment by mind04** [comment]
NSEC3 was fixed in #9873

**Comment by monodhs** [comment]
> We do provide packages for buster on our repository: https://repo.powerdns.com/

Got 'ny Debian buster packages for architectures other than x86?

ARM64 is what we run on over here.

**Comment by cmouse** [comment]
> 
> 
> > We do provide packages for buster on our repository: https://repo.powerdns.com/
> 
> Got 'ny Debian buster packages for architectures other than x86?
> 
> ARM64 is what we run on over here.

Did you bother to open the page to peer into the raspbian section?

**Comment by monodhs** [comment]
> > > We do provide packages for buster on our repository: https://repo.powerdns.com/
> > 
> > 
> > Got 'ny Debian buster packages for architectures other than x86?
> > ARM64 is what we run on over here.
> 
> Did you bother to open the page to peer into the raspbian section?

No, because I was looking for _Debian 10 (Buster)_, specifically **and** exclusively.

Please explain why I would bother to look for something else, if _Debian 10 (Buster)_ is the **only** distribution I am interested in.

**Comment by Habbie** [comment]
> Got 'ny Debian buster packages for architectures other than x86?

not yet - tracked in #8655. If you have docker, they are easy to build using `builder/build.sh`.

**Comment by monodhs** [comment]
I can finally confirm that setting `direct-dnskey=no` solves the problem of omitted DNSKEY bits in NSEC3 RRs served in response to AXFRs, with pre-signed zones.

Given this welcome discovery, I am much more confused about the purpose of the `direct-dnskey` parameter. What are/were the reasons for its existence, since, obviously, DNSKEY RRs in master files are accepted regardless of it being set or unset?

Thanks!

**Comment by klaus-nicat** [comment]
For pre-signed slave.zones I do not see a reason for direct-dnskey=no - it should be "yes" implictely.

For unsigned slave zones, which are signed by powerdns, or for master zones which are signed by powerdns, this feature allows to:
- have DNSKEYs (for signing) in cryptkeys table and DNSKEY (for publishing) in records table (direct-dnskey=yes)
-  have DNSKEYs (for signing) in cryptkeys table only (direct-dnskey=no)

Honestly, I do not see a use case for "direct-dnskey=no".

**Comment by monodhs** [comment]
> Honestly, I do not see a use case for "direct-dnskey=no".

I am using PowerDNS to operate a hidden primary master, which I then load with the pre-signed zone data from [master files](https://datatracker.ietf.org/doc/html/rfc1035#section-5), as it were, because that seemed to be the approach with the lowest barrier of entry for setting up a pre-signed zone-only regime. That makes them pre-signed master zones, according to your taxonomy, I guess.

However, it appears that PowerDNS, because of it insisting to generate NSEC3 RRs and even sign itself, unless explicitly told not to, even if they are already provided by the zone's data source, isn't the best tool to accomplish that goal in a scalable fashion.

The pre-signing requirement derives from a strict digital sovereignty policy, which rejects collusion of serving and signing, two very different processes, and which prevents many of the attacks [discussed](https://datatracker.ietf.org/doc/html/rfc6394#section-3.3) in the context of DANE.

Be that as it may, the documentation to `direct-dnskey` is misleading, when it talks about _zonefiles_ specifically, only to [apply it](https://doc.powerdns.com/authoritative/dnssec/migration.html#secure-transfers) to other backends such as those for databases. Notice that I allow myself to make the distinction between _file_ and _database_.

Thanks!

**Comment by Habbie** [comment]
Hello! As far as I can see, this problem has been fixed in 4.4. If you think our docs are unclear in any way (I haven't been able to discern where), we'd love to receive PRs improving the situation.

I'm closing this ticket.
